django/db/models/sql/query.py

```diff
@@ -1608,10 +1608,30 @@ def trim_joins(self, targets, joins, path):
             self.unref_alias(joins.pop())
         return targets, joins[-1], joins
 
+    def resolve_cols(self, name):
+        cols = []
+
+        def _resolve_cols(expressions):
+            for exp in expressions:
+                if not isinstance(exp, Col):
+                    return _resolve_cols(exp.get_source_expressions())
+                else:
+                    cols.append(exp.alias)
+        _resolve_cols(self.annotations[name].get_source_expressions())
+        return set(cols)
+
     def resolve_ref(self, name, allow_joins=True, reuse=None, summarize=False, simple_col=False):
         if not allow_joins and LOOKUP_SEP in name:
             raise FieldError("Joined field references are not permitted in this query")
         if name in self.annotations:
+            if not allow_joins and self.annotations[name].contains_column_references:
```

Comment:
You can drop the `.contains_column_references` as it's wasteful in this case because of the way it walks the expression tree.